File: MockLibAndroid/SmartMockLib/src/main/java/com/crescentflare/smartmock/responsegenerator/SmartMockEndPointFinder.java
package com.crescentflare.smartmock.responsegenerator;

import android.content.Context;

import com.crescentflare.smartmock.utility.SmartMockFileUtility;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.InputStream;

/**
 * Smart mock library response generator: find the end point based on the path (with wildcard matching)
 */
public class SmartMockEndPointFinder
{
    /**
     * Private constructor, only static methods allowed
     */

    private SmartMockEndPointFinder()
    {
    }


    /**
     * Utility functions
     */

    private static String findFileServerPath(Context context, String path)
    {
        int slashIndex = path.lastIndexOf('/');
        while (slashIndex >= 0)
        {
            path = path.substring(0, slashIndex);
            InputStream inputStream = SmartMockFileUtility.open(context, path + "/properties.json");
            if (inputStream != null)
            {
                String jsonString = SmartMockFileUtility.readFromInputStream(inputStream);
                JSONObject jsonObject = new JSONObject();
                try
                {
                    jsonObject = new JSONObject(jsonString);
                }
                catch (JSONException ignored)
                {
                }
                String generates = jsonObject.optString("generates", "");
                if (generates.equals("fileList"))
                {
                    return path;
                }
            }
            if (path.endsWith("//"))
            {
                break;
            }
            slashIndex = path.lastIndexOf('/');
        }
        return null;
    }

    public static String findLocation(Context context, String rootPath, String requestPath)
    {
        // Return early if request path is empty
        if (requestPath.isEmpty() || requestPath.equals("/"))
        {
            return rootPath;
        }

        // Determine path to traverse
        if (requestPath.charAt(0) == '/')
        {
            requestPath = requestPath.substring(1);
        }
        if (requestPath.charAt(requestPath.length() - 1) == '/')
        {
            requestPath = requestPath.substring(0, requestPath.length() - 1);
        }
        String[] pathComponents = requestPath.split("/");

        // Start going through the file tree until a path is found
        if (pathComponents.length > 0)
        {
            String checkPath = rootPath;
            for (int i = 0; i < pathComponents.length; i++)
            {
                String pathComponent = pathComponents[i];
                String[] fileList = SmartMockFileUtility.list(context, checkPath);
                if (stringArrayContains(fileList, pathComponent))
                {
                    boolean isFile = false;
                    if (i + 1 == pathComponents.length)
                    {
                        isFile = SmartMockFileUtility.getLength(context, checkPath + "/" + pathComponent) > 0;
                    }
                    checkPath += "/" + pathComponent;
                    if (isFile)
                    {
                        String fileServerPath = findFileServerPath(context, checkPath);
                        if (fileServerPath != null)
                        {
                            checkPath = fileServerPath;
                        }
                    }
                }
                else if (stringArrayContains(fileList, "any"))
                {
                    checkPath += "/any";
                }
                else
                {
                    return null;
                }
            }
            return checkPath;
        }
        return rootPath;
    }

    private static boolean stringArrayContains(String[] stringArray, String string)
    {
        for (String check : stringArray)
        {
            if (check.equals(string))
            {
                return true;
            }
        }
        return false;
    }
}

File: MockLibIOS/SmartMockLib/Classes/responsegenerator/SmartMockEndPointFinder.swift
//
//  SmartMockEndPointFinder.swift
//  SmartMockServer Pod
//
//  Main library response generator: find the end point based on the path (with wildcard matching)
//

class SmartMockEndPointFinder {
    
    // --
    // MARK: Initialization
    // --
    
    private init() {
        // Private constructor, only static methods allowed
    }


    // --
    // MARK: Utility functions
    // --
    
    private static func findFileServerPath(path: String) -> String? {
        var traversePath = path
        while let slashIndex = traversePath.range(of: "/", options: .backwards)?.lowerBound {
            traversePath = traversePath.substring(to: slashIndex)
            if let inputStream = SmartMockFileUtility.open(path: traversePath + "/properties.json") {
                let jsonString = SmartMockFileUtility.readFromInputStream(inputStream)
                if let jsonDict = SmartMockStringUtility.parseToDictionary(string: jsonString) {
                    if jsonDict["generates"] as? String == "fileList" {
                        return traversePath
                    }
                }
            }
            if traversePath.hasSuffix("//") {
                break
            }
        }
        return nil
    }
    
    static func findLocation(atRootPath: String, checkRequestPath: String) -> String? {
        // Return early if request path is empty
        if checkRequestPath.isEmpty || checkRequestPath == "/" {
            return atRootPath
        }

        // Determine path to traverse
        var requestPath = checkRequestPath
        if requestPath.characters[requestPath.startIndex] == "/" {
            requestPath = requestPath.substring(from: requestPath.characters.index(requestPath.startIndex, offsetBy: 1))
        }
        if requestPath.characters[requestPath.characters.index(requestPath.startIndex, offsetBy: requestPath.characters.count - 1)] == "/" {
            requestPath = requestPath.substring(to: requestPath.characters.index(requestPath.startIndex, offsetBy: requestPath.characters.count - 1))
        }
        let pathComponents = requestPath.characters.split{ $0 == "/" }.map(String.init)
        
        // Start going through the file tree until a path is found
        if pathComponents.count > 0 {
            var checkPath = atRootPath
            for i in 0..<pathComponents.count {
                let pathComponent: String = pathComponents[i]
                if let fileList = SmartMockFileUtility.list(fromPath: checkPath) {
                    if fileList.contains(pathComponent) {
                        var isFile = false
                        if i + 1 == pathComponents.count {
                            isFile = SmartMockFileUtility.getLength(ofPath: checkPath + "/" + pathComponent) > 0
                        }
                        checkPath += "/" + pathComponent
                        if isFile {
                            if let fileServerPath = findFileServerPath(path: checkPath) {
                                checkPath = fileServerPath
                            }
                        }
                    } else if fileList.contains("any") {
                        checkPath += "/any"
                    } else {
                        return nil
                    }
                } else {
                    return nil
                }
            }
            return checkPath
        }
        return atRootPath
    }
    
}
